server/mqtt.go

```diff
@@ -2833,6 +2960,30 @@ func (as *mqttAccountSessionManager) transferRetainedToPerKeySubjectStream(log *
 	return nil
 }
 
+func (as *mqttAccountSessionManager) getCachedRetainedMsg(subject string) *mqttRetainedMsg {
+	v, ok := as.rmsCache.Load(subject)
+	if !ok {
+		return nil
+	}
+	rm := v.(mqttRetainedMsg)
+	if rm.expiresFromCache.Before(time.Now()) {
+		as.rmsCache.Delete(subject)
+		return nil
+	}
+	return &rm
+}
+
+func (as *mqttAccountSessionManager) setCachedRetainedMsg(subject string, rm mqttRetainedMsg, onlyReplace bool) {
```

Comment:
The fact that `rm` is not a pointer causes a copy of the whole struct when this function is invoked. So the caller is already doing `setCachedRetainedMsg(key, *rm, true)`, then a copy is made by the runtime to pass as a parameter. Does not seem too efficient. Also here it is actually what you want, but modifying `rm` would not modify the `rm` that is passed to the function.

Reply:
A bit of a mess up on my part, the intention was to avoid storing the (extra) pointers in the map, but it is simply not worth the added complexity. Storing (and passing) pointers now